src/cpp/include/utils/vision_pipeline.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include <opencv2/opencv.hpp>
+#include <openvino/openvino.hpp>
+
+#include "adapters/inference_adapter.h"
+#include "tasks/results.h"
+#include "utils/tiling.h"
+
+template <typename ResultType>
+class Pipeline {
+public:
+    Pipeline() {}
+    virtual ResultType infer(cv::Mat image) = 0;
+    virtual void setCallback(std::function<void(ResultType, ov::AnyMap)> callback) = 0;
+    virtual void inferAsync(cv::Mat image, ov::AnyMap user_data) = 0;
+    virtual std::vector<ResultType> inferBatch(std::vector<cv::Mat> images) = 0;
+};
+
+template <typename ResultType>
+class VisionPipeline : public Pipeline<ResultType> {
+private:
+    std::shared_ptr<InferenceAdapter> adapter;
+    std::function<InferenceInput(cv::Mat)> preprocess;
+    std::function<ResultType(InferenceResult)> postprocess;
+
+public:
+    VisionPipeline() {}
+    VisionPipeline(std::shared_ptr<InferenceAdapter> adapter,
+                   std::function<InferenceInput(cv::Mat)> preprocess,
+                   std::function<ResultType(InferenceResult)> postprocess)
+        : adapter(adapter),
+          preprocess(preprocess),
+          postprocess(postprocess) {}
+
+    inline ResultType infer(cv::Mat image) {
+        auto input = preprocess(image);
+        InferenceResult result;
+        result.inputImageSize = image.size();
+        result.data = adapter->infer(input);
+        return postprocess(result);
+    }
+
+    inline void setCallback(std::function<void(ResultType, ov::AnyMap)> callback) {
+        adapter->setCallback([&](ov::InferRequest request, CallbackData additional_data) {
+            InferenceResult result;
+            size_t index = additional_data->at("index").as<size_t>();
+            result.inputImageSize = additional_data->at("inputImageSize").as<cv::Size>();
+            auto user_data = additional_data->at("user_data").as<ov::AnyMap>();
+            for (const auto& item : adapter->getOutputNames()) {
+                result.data.emplace(item, request.get_tensor(item));
+            }
+            callback(postprocess(result), user_data);
+        });
+    }
+
+    inline void inferAsync(cv::Mat image, ov::AnyMap user_data) {
+        auto additional_data = std::make_shared<ov::AnyMap>();
+        additional_data->insert({"inputImageSize", image.size()});
+        additional_data->insert({"user_data", user_data});
+        adapter->inferAsync(preprocess(image), additional_data);
+    }
+
+    inline std::vector<ResultType> inferBatch(std::vector<cv::Mat> images) {
+        auto results = std::vector<ResultType>(images.size());
```

Comment:
Looking at the implementation, it seems beneficial to delegate inferBatch to the adapter somehow, as some of the backends can support real batch inference, so having it's async emulation looks as a constraint of the pipeline.